[Woo POS] Make reader alert types hashable and identifiable in preparation for posModal changes

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #13234
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds Hashable and Identifiable conformance to `PointOfSaleCardPresentPaymentAlertType`, to allow it to be used in a future `Item?` binding for presentation of the posModal.

This is preparation, and it doesn't change any user-facing functionality, but it's worth checking the modals still work as expected.

There are a few manual conformances to `Equatable` and `Hashable` required – primarily because they have closures and Images, which can't have synthesised functions for these protocols. I don't love this, and I've tried to avoid it as much as possible by using `imageName` instead of an Image in the view model, but on software updates it wasn't practical to do much else.

Since we don't have good reflection, my testing of these conformances is limited. It seemed pointless to write property-by-property tests, instead I opted for tests which will act as a reminder if we add a property, to update the conformance.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app and navigate to Point of Sale
2. Turn off bluetooth
3. Tap `Connect your reader`
4. Observe that the error is shown correctly
5. Turn on Bluetooth and try again
6. Turn off your reader as soon as it starts connecting
7. Observe that the error is shown correctly
8. Turn on your reader, tap cancel, and then try again (note – there's an underlying issue with payments where it won't find the same reader if you tap `Try again` in this case.)
9. Observe that the connection success is shown correctly

### Updates and low battery

1. Enable the simulated reader
2. Set the `Terminal.shared.simulatorConfiguration.availableReaderUpdate = .lowBattery` in the `StripeCardReaderService`
3. Launch the app and navigate to POS mode
4. Connect your reader
5. Observe that the error shows correctly

Repeat with the update set to `.required` – observe that the update screens show correctly.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I've tested the above scenarios on an iPad running iOS 16, and a simulated iPad running iOS 17

I've tested the existing app's equivalent flows on an iPad and an iPhone.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/208bc2fe-a80a-4153-a232-c7b7362261ff


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
13234 Use imageName in connection view models

This will help us implement hashable and react to view changes more effectively.

joshheald committed:
Make alert type hashable to try to get them to update

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/219">20.1</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
13234 Make card reader alert types identifiable

joshheald committed:
13234 Test alert type equatable conformance

It’s not a great test, but reflection is so limited it’s the best I could do. At least it should be a reminder to update the `==` function when we change these.

joshheald committed:
13234 Add manual hashable conformance tests

These are almost all identical to the equatable tests… but having the second one there is a good reminder to future-us when we change these view models.

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/Card Present Payments/Connection Alerts/PointOfSaleCardPresentPaymentConnectingFailedUpdateAddressAlertViewModel.swift:
```diff
@@ -51,6 +51,31 @@ final class PointOfSaleCardPresentPaymentConnectingFailedUpdateAddressAlertViewM
     }
 }
 
+extension PointOfSaleCardPresentPaymentConnectingFailedUpdateAddressAlertViewModel: Hashable {
```
We could simplify this conformance by having an unique id and relying on that, but I understand that's not what we want, right? We do want to be able to compare all properties? 

Just wondering: What about simplifying some of its properties for the conformance? Like using `cancelButtonViewModel.id` rather than the whole `cancelButtonViewModel`?

joshheald replied:
The `hashValue` of a Hashable object should change when the (essential components of the) object change. For synthesised Hashable, that's whenever any property changes, and we should aim to get as close to that as possible.

> Hashing a value means feeding its essential components into a hash function, represented by the Hasher type. Essential components are those that contribute to the type’s implementation of Equatable.
> [...]
> In your hash(into:) implementation, call combine(_:) on the provided Hasher instance with the essential components of your type. To ensure that your type meets the semantic requirements of the Hashable and Equatable protocols, it’s a good idea to also customize your type’s Equatable conformance to match.
> – [Conforming to the Hashable protocol (Apple docs)](https://developer.apple.com/documentation/swift/hashable#Conforming-to-the-Hashable-Protocol)

Using (only) random IDs for hash values would lead to an incorrect implementation. 

Take the scenario where we make an object as follows:

```
struct Thing: Identifiable, Equatable {
    let id: String
    var title: String
}

extension Thing: Hashable {
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}
```

Ignore that Hashable could be synthesised for this struct – we've written our own.

```
var foo = Thing(id: "123", title: "An awesome title")
let initialFoo = foo
let initialHash = foo.hashValue
foo.title = "Short title"
let laterHash = foo.hashValue
XCTAssertNotEqual(initialHash, laterHash) // Fails, because the hash only changes for a new `id`, but should pass with correct hashable implementation.
XCTAssertNotEqual(initialFoo, foo) // Passes, because equatable works fine
```

In SwiftUI, a change in the hash value gets used to determine whether the object has changed. Good objects to observe in SwiftUI will have a consistent ID (so we know it's the same object, and don't think the changed object is new) but a changing hashValue (so we know the object has updates and should be redrawn)

This comes in to play visually with animation and lists, but I think it also matters for deeper performance reasons in SwiftUI.

I'm going to double check my equatable implementations because I think there was one strange case where I was able to rely on the synthesised Equatable, but had to write a custom Hashable. (PointOfSaleCardPresentPaymentReaderUpdateCompletionAlertViewModel)
Edit: done in f3f8a6e

joshheald replied:
I don't like that we can't compare closure properties for the purposes of hashable.

It helps that they're all let constants, but I'm going to make sure the structs with closures all have `let id = UUID()` as well, so that we can't end up with two different copies that have different actions but the same hashvalue.

If we needed to have `var` closures, one solution would be to have a private property which has a UUID or random value, which we change whenever the closure property is set and compare in the equatable/hashable functions.

Edit: done in 20a41c6

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/Card Present Payments/Connection Alerts/PointOfSaleCardPresentPaymentConnectingFailedAlertViewModel.swift:
```diff
@@ -2,9 +2,9 @@ import Foundation
+struct PointOfSaleCardPresentPaymentConnectingFailedAlertViewModel: Hashable {
     let title = Localization.title
-    let image = Image(uiImage: .paymentErrorImage)
+    let imageName = PointOfSaleAssets.paymentsError.imageName
```
Really nice the way we remove specific presentation datatypes from the view model and just rely on simpler ones, in this case (and many others, but not all of them) we can also remove the `import SwiftUI` bit if needed.

joshheald replied:
Done in 42304e9

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/Card Present Payments/Connection Alerts/PointOfSaleCardPresentPaymentFoundMultipleReadersAlertViewModel.swift:
```diff
@@ -15,3 +15,14 @@ struct PointOfSaleCardPresentPaymentFoundMultipleReadersAlertViewModel {
+extension PointOfSaleCardPresentPaymentFoundMultipleReadersAlertViewModel: Hashable {
+    static func == (lhs: PointOfSaleCardPresentPaymentFoundMultipleReadersAlertViewModel,
+                    rhs: PointOfSaleCardPresentPaymentFoundMultipleReadersAlertViewModel) -> Bool {
+        return lhs.readerIDs == rhs.readerIDs
```
Question: Possibly doesn't matter at all but since `readerIDs` it's an array of Strings, should we consider equal 2 arrays with the same values but in different order? Technically these are different, but in our case if contains the same reader id's then should they be the same?

joshheald replied:
They are different, and we'd want equatable to return false (which it does). If the order of the array changes, so should a sort order on screen, for example.

iamgabrielma commented on WooCommerce/WooCommerceTests/Testing/XCTestCase+PropertyCount.swift:
```diff
@@ -0,0 +1,23 @@
+                                   messageHint: String? = nil,
+                                   file: StaticString = #filePath,
+                                   line: UInt = #line) {
+        let mirror = Mirror(reflecting: instance)
```
This is something I didn't know it existed, made me go through some TIL rabbit hole. Nice!

joshheald replied:
Yeah me too, I kinda knew there was some basic reflection added to Swift a while ago, but I'd never used it as it is so limited. This is kind of helpful though.

iamgabrielma reviewed (approved):
Works well! 

Something unrelated to the changes: When running the simulator, on `.lowBattery`, I get an empty list of "readers found": 

![Simulator Screen Recording - iPad Pro (12 9-inch) (6th generation) - 2024-08-23 at 11 42 52](https://github.com/user-attachments/assets/772b5918-c213-46d6-a427-51f79de23d8d)

joshheald committed:
13234 Remove unused SwiftUI imports

joshheald committed:
13234 Make custom `hash` and `==` functions match

joshheald committed:
13234 Use IDs to compare structs with closures

Closures can’t be compared, which makes it difficult to identify differences between objects which hold them as properties.

Where they can’t be changed, including a pseudo-random ID on an object is a helpful tool to stop us from saying that two buttons with the same title but different actions are equal or have the same hash.

If the closures are held in a `var`, we would need to instead update a random identifier (*not* the object’s ID) every time they were set. The random identifier can then act as a proxy for that closure in comparison operations.

It’s still not perfect, but it’s safer.
